### ISSUE TITLE
Consider adding `deprecated_consistency` to recommended

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/lints#60]
It's frustrating to remove something that you think was properly deprecated only to discover that many users of it never saw the deprecation warning. Enabled in Flutter.

**Comment by mit-mit** [comment]
This seems like a tooling issue: Any use of a member or constructor of a class that is marked deprecated should be flagged.

**Comment by mit-mit** [comment]
@mit-mit to file tools issue

**Comment by mit-mit** [comment]
@goderbauer do you have an example of a case where you are not seeing a warning?

I tried to create a class marked deprecated, and I get usage warnings on both static method calls and on constructor calls.

**Comment by goderbauer** [comment]
@Piinks Do you remember where we saw the inconsistencies around deprecation that let us to adopting the `deprecated_consistency` lint?

**Comment by mit-mit** [comment]
cc @pq, is that fixable?

**Comment by bwilkerson** [comment]
Just to clarify, I think what you're asking is whether, at least for the purposes of deprecation warnings, we can consider fields and field initializing parameters to be equivalent. That is, if either is deprecated then we consider both to be deprecated.

Personally, I don't have any problem with that, but we might want to update the documentation for `deprecated` to make this association clear. @lrhn

**Comment by devoncarew** [comment]
This has been approved for the `recommended` set.

**Comment by devoncarew** [comment]
An update - this is still approved, but we'll delay adding this to the lint set until after we have a fix for https://github.com/dart-lang/linter/issues/4752#issuecomment-1728213638 (and that fix has shipped in an sdk).

**Comment by devoncarew** [comment]
@pq @lrhn - from the discussion here: https://github.com/dart-lang/linter/issues/4752#issuecomment-1729426585, are we considering this lint not necessary - that we should bake a version of `deprecated_consistency` into the warnings that we normally issues for deprecated members?

**Comment by bwilkerson** [comment]
> We should make sure to warn about unnecessary `//ignores`!

I have a CL that introduces such a check (first uploaded in March of 2021), that has not yet been able to land for various reasons that I won't go into here. Suffice it to say that warning about unnecessary ignores presents some social challenges. (Happy to talk about it more, but maybe in a different thread.)

> This is a lint about two (or more) places in the code, so we should be clear on how they interact, including where `// ignores` can go.
> ...
> Then we should make sure that the `// ignores` work as intended.

I'm not saying that we couldn't enhance the current implementation to support what you're proposing, but the implementation we have today requires that the ignore comment be on either the same line as the beginning of the diagnostic's highlight region, or by itself on the line before the beginning of the highlight region.

**Comment by lrhn** [comment]
> ..., that has not yet been able to land for various reasons that I won't go into here. 

Let's hope we can do something. :)

> ...  the implementation we have today requires that the ignore comment be on either the same line as the beginning of the diagnostic's highlight region, or by itself on the line before the beginning of the highlight region.

So commenting the `@override` is out. (I'm assuming it's not part of the region. If it was, things would get much weirder than they are today.)

This diagnostic relates to two places in the code, the variable declaration and the initializing formal, being out of sync. I'm assuming we only make one diagnostic for it (or if we make two, we'll need two `ignore`s).

It's probably better to always report it at the initializing formal, because there can be more than one of those,
which means you can, and must, ignore each one of those independently. Annoying, but doable. The assumption is that's it's going to be rare to need an ignore, and in those cases, the first few steps of the necessary refactoring will likely remove the warning, leaving only a deprecated getter backed by something else. So not a big issue.

An initializing formal is usually not that long, and easy to break onto a line of its own, so it should be easy to add a comment to that line, even if it might mean reformatting parameter list to give it a new line.
So the analysis server can insert a suitable comment, which won't be moved by dartfmt. That's probably good enough.

SGTM.

Aka:
> If an initializing formal is deprecated, and the instance variable it's initializing is not, or vice versa,
> the deprecation is considered inconsistent, and a warning is emitted for the initializing formal.
> 
> If the design is deliberate, and one of the two will be removed while the other stays in the API, 
> an `// ignore: deprecated_consistency` comment should be added to the initializing formal, 
> either at the end of the line where the initializing format starts, or alone on the line before.
>
> Using an `// ignore: deprecated_consistency` comment is the *recommended* way to signal
> that the initializing formal and instance variable are not both going away.
> The comment will become unnecessary as soon as the deprecated member has been removed,
> and possibly earlier if the class is refactored to prepare for removing the deprecated declaration.
>
> Example:
> ```dart
> class Example {
>   @deprecated
>   final int goingAway;
> 
>   final int hereToStay;
> 
>   Example.v1(
>       // ignore: deprecated_consistency
>       this.goingAway,
>       @deprecated
>       // ignore: deprecated_consistency
>       this.hereToStay);
> 
>   Example.v2(
>       this.goingAway, // ignore: deprecated_consistency
>       @deprecated this.hereToStay // ignore: deprecated_consistency
>       );
> 
>   // No warnings, everything is consistent.
>   Example.v3(@deprecated this.goingAway, this.hereToStay);
> }
> ```

This should be enough. Again, it's not going to be used often, it's likely that the comment goes away before the deprecated declaration, if any refactoring is done in anticipation. 
All in all, a short-lived `// ignore` to opt out of a warning for something which otherwise has a good chance of being a mistake, and it's easy to remove the need for the ignore (just change the initializing formal to a normal parameter, and initialize in the initializer list, if it's that important).

**Comment by bwilkerson** [comment]
> I'm assuming we only make one diagnostic for it (or if we make two, we'll need two ignores).

Correct. Given
```dart
class B {
  @Deprecated('soon')
  Object? field;

  B({this.field});
}
```
there is a single diagnostic and the highlight covers `this.field`.

> It's probably better to always report it at the initializing formal, because there can be more than one of those ...

That was our reasoning too.

> An initializing formal is usually not that long, and easy to break onto a line of its own, so it should be easy to add a comment to that line, even if it might mean reformatting parameter list to give it a new line.

While that's possible, it isn't necessary. Either of the following solutions will also work:
```dart
class B {
  @Deprecated('soon')
  Object? field;

  // ignore: deprecated_consistency
  B({this.field});
}
```
or
```dart
class B {
  @Deprecated('soon')
  Object? field;

  B({this.field}); // ignore: deprecated_consistency
}
```

**Comment by devoncarew** [comment]
Given the relative value we get from this lint, and the work necessary to fix it, we're unlikely to add this to our recommended lint set. Closing.
